spring/boot-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaHttpClientResponseSubscriber.java

```diff
@@ -42,78 +42,32 @@
  * by calling {@link #toResponseBodyPublisher()} which returns {@link ResponseBodyPublisher}.
  */
 final class ArmeriaHttpClientResponseSubscriber
-        implements Subscriber<HttpObject>, BiFunction<Void, Throwable, Void> {
+        implements BiFunction<Void, Throwable, Void> {
 
     private final CompletableFuture<ResponseHeaders> future = new CompletableFuture<>();
     private final EventLoop eventLoop = CommonPools.workerGroup().next();
+    private final EmitterProcessor<HttpData> httpDataProcessor = EmitterProcessor.create(1);
 
-    @Nullable
-    private ResponseBodyPublisher publisher;
-
-    @Nullable
-    private Subscription subscription;
+    ArmeriaHttpClientResponseSubscriber(HttpResponse httpResponse) {
+        final EmitterProcessor<HttpObject> httpObjectProcessor = EmitterProcessor.create(1);
+        httpObjectProcessor
+                .ofType(HttpData.class)
+                .subscribe(httpDataProcessor);
 
-    private boolean isCompleted;
-    @Nullable
-    private Throwable completedCause;
+        httpObjectProcessor
+                .ofType(ResponseHeaders.class)
```

Comment:
question: does this `httpObjectProcessor` relays the trailers which is `HttpHeaders` as well?

Reply:
`ArmeriaHttpClientResponseSubscriber`'s procedure is as follows.

1. `ArmeriaHttpClientResponseSubscriber` subscribes `HttpResponse`.
2. Receive an `HttpHeaders` from response of `1`
3. `Subscriber` completes `httpHeadersFuture`.
4. `ArmeriaClientHttpResponse` subscribes `ResponseBodyPublisher` to create a repsonse body flux.

If `ArmeriaHttpClientResponseSubscriber` doesn't use `EmitterProcessor` for HttpData, at the moment of third step all httpObjectProcessor's subscribers cancel their subscirptions. httpObjectProcessor also cancels it's subscription from `HttpResponse` and remembers itself completed. After that, `ArmeriaClientHttpResponse` subscribes `ResponseBodyPublisher` using `httpObjectProcessor`. It will not receive HttpData, and it will receive an only completed event.

If `ArmeriaHttpClientResponseSubscriber` uses `EmitterProcessor` for HttpData, at the moment of third step subscription of header is canceled. But subscription of HttpData still remains. And `httpObjectProcessor` also remains it's subscription. Therefore it can send all of httpData to `ArmeriaClientHttpResponse`.